File: planka-ui/src/types/builtin-field.ts
/**
 * 内置字段定义
 */
export interface BuiltinFieldDef {
  /** 字段ID，以 $ 开头 */
  id: string
  /** 显示名称 */
  name: string
  /** 字段类型 */
  fieldType: string
  /** 是否可排序 */
  sortable: boolean
  /** 是否可编辑 */
  editable: boolean
}

/**
 * 内置字段列表
 */
export const BUILTIN_FIELDS: BuiltinFieldDef[] = [
  { id: '$createdAt', name: '创建时间', fieldType: 'DATE', sortable: true, editable: false },
  { id: '$updatedAt', name: '更新时间', fieldType: 'DATE', sortable: true, editable: false },
  { id: '$statusId', name: '价值流状态', fieldType: 'ENUM', sortable: false, editable: true },
  { id: '$cardStyle', name: '卡片周期', fieldType: 'ENUM', sortable: true, editable: false },
  { id: '$archivedAt', name: '归档时间', fieldType: 'DATE', sortable: true, editable: false },
  { id: '$discardedAt', name: '丢弃时间', fieldType: 'DATE', sortable: true, editable: false },
  { id: '$code', name: '卡片编号', fieldType: 'TEXT', sortable: true, editable: false },
  { id: '$description', name: '详情描述', fieldType: 'MARKDOWN', sortable: false, editable: true },
]

/**
 * 判断是否为内置字段
 * @param fieldId 字段ID
 * @returns 如果以 $ 开头则为内置字段
 */
export function isBuiltinField(fieldId: string): boolean {
  return fieldId != null && fieldId.startsWith('$')
}

/**
 * 获取内置字段定义
 * @param fieldId 字段ID
 * @returns 内置字段定义，未找到时返回 undefined
 */
export function getBuiltinField(fieldId: string): BuiltinFieldDef | undefined {
  return BUILTIN_FIELDS.find((f) => f.id === fieldId)
}

/**
 * 卡片周期状态映射
 */
export const CARD_STYLE_LABELS: Record<string, string> = {
  ACTIVE: `活跃`,
  ARCHIVED: '已归档',
  DISCARDED: '已丢弃',
}

/**
 * 卡片周期状态颜色
 */
export const CARD_STYLE_COLORS: Record<string, string> = {
  ACTIVE: '#52c41a',
  ARCHIVED: '#8c8c8c',
  DISCARDED: '#d4380d',
}

/**
 * 价值流阶段类型颜色（浅色系，适配黑色文字）
 */
export const STEP_KIND_COLORS: Record<string, string> = {
  TODO: '#ffe27a',        // 待办 - 黄色
  IN_PROGRESS: '#c8d9ff', // 进行中 - 蓝色
  DONE: '#b4db94',        // 完成 - 绿色
  CANCELLED: '#d9d9d9',   // 取消 - 灰色
}

/**
 * 获取阶段类型对应的颜色
 * @param stepKind 阶段类型
 * @returns 颜色值，默认返回浅蓝色
 */
export function getStepKindColor(stepKind?: string): string {
  const defaultColor = STEP_KIND_COLORS.IN_PROGRESS ?? '#c8d9ff'
  if (!stepKind) return defaultColor
  return STEP_KIND_COLORS[stepKind] ?? defaultColor
}

File: planka-kernel/planka-domain/src/main/java/dev/planka/domain/field/BuiltinField.java
package dev.planka.domain.field;

import dev.planka.domain.schema.definition.fieldconfig.FieldType;
import lombok.Getter;

/**
 * 内置字段枚举
 * <p>
 * 定义卡片的系统内置属性，用于视图列配置。
 * 内置字段ID使用 $ 前缀与自定义属性区分。
 */
@Getter
public enum BuiltinField {

    /**
     * 创建时间
     */
    CREATED_AT("$createdAt", "创建时间", FieldType.DATE, true, false),

    /**
     * 更新时间
     */
    UPDATED_AT("$updatedAt", "更新时间", FieldType.DATE, true, false),

    /**
     * 价值流状态
     */
    STATUS_ID("$statusId", "价值流状态", FieldType.ENUM, true, true),

    /**
     * 卡片周期（活跃/已归档/已丢弃）
     */
    CARD_STYLE("$cardStyle", "卡片周期", FieldType.ENUM, true, false),

    /**
     * 归档时间
     */
    ARCHIVED_AT("$archivedAt", "归档时间", FieldType.DATE, true, false),

    /**
     * 丢弃时间
     */
    DISCARDED_AT("$discardedAt", "丢弃时间", FieldType.DATE, true, false),

    /**
     * 卡片编号（优先 customCode，其次 codeInOrg）
     */
    CODE("$code", "卡片编号", FieldType.TEXT, true, false);

    /**
     * 字段ID，以 $ 开头
     */
    private final String fieldId;

    /**
     * 显示名称
     */
    private final String displayName;

    /**
     * 字段类型
     */
    private final FieldType fieldType;

    /**
     * 是否可排序
     */
    private final boolean sortable;

    /**
     * 是否可编辑
     */
    private final boolean editable;

    BuiltinField(String fieldId, String displayName, FieldType fieldType, boolean sortable, boolean editable) {
        this.fieldId = fieldId;
        this.displayName = displayName;
        this.fieldType = fieldType;
        this.sortable = sortable;
        this.editable = editable;
    }

    /**
     * 判断是否为内置字段ID
     *
     * @param fieldId 字段ID
     * @return 如果以 $ 开头则为内置字段
     */
    public static boolean isBuiltinField(String fieldId) {
        return fieldId != null && fieldId.startsWith("$");
    }

    /**
     * 根据字段ID查找内置字段
     *
     * @param fieldId 字段ID
     * @return 对应的内置字段枚举值，未找到时返回 null
     */
    public static BuiltinField fromFieldId(String fieldId) {
        if (fieldId == null) {
            return null;
        }
        for (BuiltinField field : values()) {
            if (field.fieldId.equals(fieldId)) {
                return field;
            }
        }
        return null;
    }
}
